git-commitgraph/src/graph/init.rs

```diff
@@ -0,0 +1,104 @@
+use crate::{graph_file, Graph, GraphFile, MAX_COMMITS};
+use git_object::HashKind;
+use quick_error::quick_error;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Corrupt(msg: String, path: PathBuf) {
+            display("{}", msg)
+        }
+        GraphFile(err: graph_file::Error, path: PathBuf) {
+            display("{}", path.display())
+            source(err)
+        }
+        HashVersionMismatch(path1: PathBuf, hash1: HashKind, path2: PathBuf, hash2: HashKind) {
+            display(
+                "Commit-graph files mismatch: '{}' uses hash {:?}, but '{}' uses hash {:?}",
+                path1.display(),
+                hash1,
+                path2.display(),
+                hash2,
+            )
+        }
+        Io(err: std::io::Error, path: PathBuf) {
+            display("Could not open commit-graph graph_file at '{}'", path.display())
+            source(err)
+        }
+        TooManyCommits(num_commits: u64) {
+            display(
+                "Commit-graph files contain {} commits altogether, but only {} commits are allowed",
+                num_commits,
+                MAX_COMMITS,
+            )
+        }
+        // This might actually be legal...
+        VersionMismatch(path1: PathBuf, version1: graph_file::Kind, path2: PathBuf, version2: graph_file::Kind) {
+            display(
+                "Commit-graph files mismatch: '{}' is version {:?}, but '{}' is version {:?}",
+                path1.display(),
+                version1,
+                path2.display(),
+                version2,
+            )
+        }
+    }
+}
+
+impl Graph {
+    pub fn from_object_dir(object_dir: impl AsRef<Path>) -> Result<Self, Error> {
+        Self::from_single_file(object_dir.as_ref()).or_else(|_| Self::from_split_chain(object_dir.as_ref()))
+    }
+
+    pub fn from_single_file(object_dir: impl AsRef<Path>) -> Result<Self, Error> {
+        let single_graph_file = object_dir.as_ref().join("info").join("commit-graph");
+        let file = GraphFile::at(&single_graph_file).map_err(|e| Error::GraphFile(e, single_graph_file.clone()))?;
+        Self::new(vec![file])
+    }
+
+    pub fn from_split_chain(object_dir: impl AsRef<Path>) -> Result<Self, Error> {
```

Comment:
With the disclaimer of having only partial knowledge of the format itself, let me share some thoughts.

I think all the init functions are a little smarter than they have to be, as the only two (or more) files we seem to be interested in is the `commit-graph*` ones in _some_ directory. Thus it would be enough to provide a directory path to a method, which will figure things out from there.

In these low-level crates I would refrain from enforcing structures usually seen in an actual git repository. This would also help with accessing fixtures in tests, which right now need nested directory structures including a pack directory even though commit-graphs (as far as I could tell) have no business with them.

Reply:
Well sometimes there is a single `objects/info/commit-graph` file and sometimes there is an `objects/info/commit-graphs` directory of files. Are you saying passing the `objects/info/` directory instead of `objects/` is preferred? I'm fine with that.

Reply:
Yes, that's what I meant. Thank you!